### PR TITLE
WIP: Improved the patch file screen

### PR DESCRIPTION
**Pull request by weyert**
My first attempt to get color coding and diffing working together. Purposely attaching a screen where it's not working perfectly. Let me know, what you think of it. 

Screenshot:
Image

### CODE REVIEW

**Comment by weyert** [comment]
@Sherlouk @BasThomas @rnystrom Played around a bit :)

File: Pods/Highlightr/Pod/Classes/HTMLUtils.swift [hunk 1]
```diff
@@ -0,0 +1,294 @@
+
+import Foundation
+
+//Adapted from: https://gist.github.com/jrmgx/3f9f1d330b295cf6b1c6
+
+private let characterEntities : [String: Character] = [
```
Comment (Sherlouk) [inline Pods/Highlightr/Pod/Classes/HTMLUtils.swift]: Does this duplicate anything from the HTMLString pod?

File: Classes/Views/DiffString.swift [hunk 2]
```diff
@@ -7,6 +7,35 @@
...
 
 import UIKit
+import Highlightr
+
+func CreateColorCodedString(code: String, includeDiff: Bool=false) -> NSAttributedString {
+    guard let highlightr = Highlightr() else { return NSAttributedString(string: code) }
+
+    if includeDiff {
+        highlightr.setTheme(to: "github")
+        let colorCodedString = highlightr.highlight(code, as: "swift") ?? NSAttributedString(string: code)
```
Comment (Sherlouk) [inline Classes/Views/DiffString.swift]: Don't think we can/should assume the language, could be anything! Objective-C 😷

File: Classes/Views/DiffString.swift [hunk 3]
```diff
@@ -7,6 +7,35 @@
...
+            if line.hasPrefix("+") {
+                attributedText.addAttribute(.backgroundColor, value: Styles.Colors.Green.light.color as Any, range: nsrange)
+            } else if line.hasPrefix("-") {
+                attributedText.addAttribute(.backgroundColor, value: Styles.Colors.Red.light.color as Any, range: nsrange)
+            }
+        }
+
+        return attributedText
+    } else {
+        highlightr.setTheme(to: "github")
```
Comment (Sherlouk) [inline Classes/Views/DiffString.swift]: Both parts of `if` statement set theme, should move this out top

File: Classes/Views/CodeTextView.swift [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+//  CodeTextView.swift
+//  Freetime
+//
+//  Created by Weyert de Boer on 12/10/2017.
+//  Copyright © 2017 Ryan Nystrom. All rights reserved.
+//
+
+import UIKit
+
+final class CodeTextView: UIView, UITextViewDelegate {
```
Comment (Sherlouk) [inline Classes/Views/CodeTextView.swift]: I'm liking where this is going, be great to see it finished!

**Comment by Sherlouk** [comment]
In reference to the image above, is it possible to get green inbetween the gaps? Feels like I'm seeing `+` for every line but green only sometimes?

Out of scope for this card, but on tablet it would be nice to get side-by-side comparison à la GitHub

File: Pods/Highlightr/Pod/Classes/HTMLUtils.swift (hunk 1 above)
Comment (weyert) [reply Pods/Highlightr/Pod/Classes/HTMLUtils.swift]: I think it does but I haven't verified it.

File: Classes/Views/DiffString.swift (hunk 2 above)
Comment (weyert) [reply Classes/Views/DiffString.swift]: You can automatically detect the language but then it didn't looked that good. See:
Image

**Comment by weyert** [comment]
@Sherlouk Yes, that's why purposely attached that image because this is the problem I am having. I am not sure why it makes the `+` on the first line green and why it sometimes goes wrong. I have to look into it. Currently the `Pod` doesn't support language + diffing together hence my current solution. This my first time I am messing around with `NSAttributedString`.

File: Classes/Views/CodeTextView.swift (hunk 4 above)
Comment (weyert) [reply Classes/Views/CodeTextView.swift]: Going to be my attempt to get line numbers. If you know a good approach don't hesitate to let me know!

**Comment by weyert** [comment]
I am not sure why it not always find the appropriate + to mark them. Anyone here having more experience with `NSAttributedString`? My other approach would be to somehow customise the HTML parsing by ourselves or maybe somehow customise the Hightltr.js so that you can combine language + diff. Anyone got suggestions? @Sherlouk @rnystrom @BasThomas ?

**Comment by Sherlouk** [comment]
https://github.com/raspu/Highlightr/blob/master/Pod/Assets/Highlighter/languages/diff.min.js

Highlightr does support a diff language?

**Comment by weyert** [comment]
Yes, but don't you want to have color coding while displaying the diff?

**Comment by Sherlouk** [comment]
image

Looking at the Highlightr demo page - The GitHub styling for diffing does have the color coding as we want:

https://highlightjs.org/static/demo/ (Select GitHub and scroll down to Diff)

**Comment by weyert** [comment]
Yes, there is a `diff` language but you can't then color code the language that's why I wrote that specific code but some lines don't get picked up.

**Comment by Sherlouk** [comment]
If you make sure colloborators can commit to your branch, I'll take a look tonight and see if I spot anything 😄

**Comment by weyert** [comment]
@Sherlouk How do I do that?

**Comment by Sherlouk** [comment]
@weyert All you need to do is tick the box on right 😄 

image

**Comment by weyert** [comment]
Oh, that's checked 👍

**Comment by Sherlouk** [comment]
Okay so the issue is this line:

```
guard let range = code.range(of: line) else { continue }
```

That's getting the range of the first instance of the provided line text. It's skipping lines like the one `//` under `// Freetime` because it's already handled a line that is exactly that. What we need to do is get a mutable version of the `code` variable. After each loop through the `lines` array, we need to remove `line` from the start of `code` ultimately getting us to the point where `code` is empty.

This will enable the `range(of: line)` to get the line we're actually looking at, rather than the first instance of it 😄

Make sense?

**Comment by weyert** [comment]
Guess, it fails when the line is not unique in the `code` variable, correct? 🤓

**Comment by Sherlouk** [comment]
Exactly

**Comment by weyert** [comment]
Okay, I will try to resolve it later today. But don't hesitate to fix it if you want. I will work on it in a hour.

**Comment by weyert** [comment]
Let's have a look!

**Comment by weyert** [comment]
@Sherlouk Working: Image

File: Classes/Utility/NSMutableAttributedString+Replace.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            if range.location != NSNotFound {
+                let lineRange = NSRange(location: range.location, length: searchLength)
+
+                switch mode {
+                case .first:
+                    self.addAttributes(attributes, range: lineRange)
+                    return
+                case .all:
+                    self.addAttributes(attributes, range: lineRange)
+                    break
```
Comment (BasThomas) [inline Classes/Utility/NSMutableAttributedString+Replace.swift]: You don't need to `break`; it's inferred.

**Comment by weyert** [comment]
I have implemented a rough version of `CodeTextView` by adapting an existing Objective-C component. Currently, it looks like this In Playground: Image

I have committed the code, could someone try if it actually compiles in the project and works? I am experiencing  sudden issues with Xcode 9 (uncaught exception) and opening projects. Restarting computer doesn't work :( If someone could have a peek, that would be awesome!

**Comment by weyert** [comment]
See the screenshot in #520. I think it's looking quite good already

**Comment by weyert** [comment]
Better version of available via pull request #577

**Pull request closed without merging**